### PR TITLE
./test: Refactoring of test script for modularization

### PR DESCRIPTION
**Pull request by ptabor**
./test: Refactoring of test script for modularization
    
This refactoring offers following benefits: 

- Unified way how go test commands are being called (in terms of flags intepretation)
- Used standard go mechanisms (like go lists) to find files/packages that are subject for test. The mechanism are module aware.
- Added instruction how to install tools needed for the tests/checkers. 
- Added colors to the output to make it easier to spot any failure. 

Confirmed to work using:
     - COVERDIR="./coverage" CPU="4" RACE=false COVER=false PASSES="build build_cov cov" ./test
     - CPU="4" RACE=false COVER=false PASSES="e2e functional integration" ./test
     - COVERDIR="./coverage" COVER="false" CPU="4" RACE="false" PASSES="fmt build unit build_cov integration e2e integration_e2e grpcproxy cov" ./test
     - PASSES=unit PKG=./wal TIMEOUT=1m ./test
     - PASSES=integration PKG=./clientv3 TIMEOUT=1m ./test
     - PASSES=unit PKG=./wal TESTCASE=TestNew TIMEOUT=1m ./test
     - PASSES=unit PKG=./wal TESTCASE="\bTestNew\b" TIMEOUT=1m ./test
     - PASSES=integration PKG=./client/integration TESTCASE="\bTestV2NoRetryEOF\b" TIMEOUT=1m ./test
     - COVERDIR=coverage PASSES="build_cov cov" ./test

### CODE REVIEW

**Comment by ptabor** [comment]
I think its ready for review. Please notice that 3419a2a is subject to review as part of https://github.com/etcd-io/etcd/pull/12319.

**Comment by ptabor** [comment]
This PR brings all the CI runs to the green status (even those that got 'allowed to fail' for a long time):

https://github.com/etcd-io/etcd/pull/12322/checks?check_run_id=1148917325

image

**Comment by xiang90** [comment]
@ptabor 

This pull request contains commits from other pending pull request it seems. so let us get others merged first.

**Comment by ptabor** [comment]
@xiang90 rebased this on top of the previous already merged PRs (thank you).

**Comment by xiang90** [comment]
@ptabor 

I took a look at the commits except the 4th one (since that one has quite some changes :P). Those are OK.

I am wondering if there is any actual change in the 4th one? Or it is just modularization work?

**Comment by ptabor** [comment]
There should be no behavioral change. But the logic is deeply refactored
E.g.
  - more standard go mechanisms to discover files - instead of "find - name
*.go"
  - flags should be more consistent
  - colored echos
  - consistent warning behavior when tool is missing
  - more comments

run_for_modules macro will allow for minimal changes when submodules are
added.

File: .travis.yml [hunk 1]
```diff
@@ -78,19 +78,22 @@ script:
           /bin/bash -c "GOARCH=amd64 PASSES='fmt bom dep' ./test"
         ;;
       linux-amd64-integration-1-cpu)
+        # TODO: Reenable 'race' when https://github.com/etcd-io/etcd/issues/12336 fixed.
         docker run --rm \
           --volume=`pwd`:/go/src/go.etcd.io/etcd gcr.io/etcd-development/etcd-test:go${TRAVIS_GO_VERSION} \
-          /bin/bash -c "GOARCH=amd64 CPU=1 PASSES='integration' ./test"
+          /bin/bash -c "GOARCH=amd64 CPU=1 PASSES='integration' RACE='false' ./test"
```
Comment (jingyih) [inline .travis.yml:84]: How flaky is `TestV3WatchRestoreSnapshotUnsync` currently? My concerns is that there might be more race conditions introduced to the codebase in the future, once we disabled race in CI testing.

File: .travis.yml [hunk 2]
```diff
@@ -30,7 +30,7 @@ env:
 matrix:
   fast_finish: true
   allow_failures:
-    - go: 1.14.3
+    - go: tip
```
Comment (jingyih) [inline .travis.yml:33]: why test against go tip for this one?

**Review by jingyih — COMMENTED** [review]
Thanks for the PR! Really love the colored outputs!

File: .travis.yml (hunk 2 above)
Comment (ptabor) [reply .travis.yml]: Reverted. I was investigating whether version of the go impacts race-detection flakiness.

File: .travis.yml (hunk 1 above)
Comment (jingyih) [reply .travis.yml:84]: Thanks! You are right. We had race detection for unit pass but not for integration.

**Comment by jingyih** [comment]
LGTM

**Comment by ptabor** [comment]
@jingyih Thank you for review. 
@jingyih @jpbetz Could you, please, merge it, as I have another PR (https://github.com/etcd-io/etcd/pull/12322) stacked on top of this one.

**Pull request merged**